src/PlonkVerify.hs

```diff
@@ -1,27 +1,172 @@
+-- provides tools for embedding existing keys from files into applications
+-- binaries.
+--
+newtype VKey = VKey
+    { _vKey :: B.ByteString }
+    deriving (Show, Eq, Ord)
+    deriving (Lift)
+
+-- | The actual proof. This is opaque cryptographic object that allows the
+-- verifier to establish that a program was executed with a particular list of
+-- public parameters in the context of the VM that is idendified with a
+-- particular verifying key.
+--
+newtype Proof = Proof
+    { _proof :: BS.ShortByteString }
+    deriving (Show, Eq, Ord)
+
+-- | A unique identifier of a program.
+--
+newtype ProgramId = ProgramId
+    { _programId :: BS.ShortByteString }
+    deriving (Show, Eq, Ord)
+
+-- | The public parameters of a program invocation.
+--
+newtype PublicParameter = PublicParameter
+    { _publicParameter :: BS.ShortByteString }
+    deriving (Show, Eq, Ord)
+
+-- | The hash of the public parameters of a program invocation.
+--
+newtype PublicParameterHash = PublicParameterHash
+    { _publicParameterHash :: BS.ShortByteString }
+    deriving (Show, Eq, Ord)
+
+mkPublicParameterHash :: BS.ShortByteString -> Either String PublicParameterHash
+mkPublicParameterHash bytes
+    | BS.length bytes /= 32 = Left $ "wrong length; expected: 32, actual: " <> show (BS.length bytes)
+    | BS.head bytes .&. 0xe0 /= 0 = Left "first three bit are not set to 0"
+    | otherwise = Right $ PublicParameterHash bytes
+
+-- | Compute the 'PublicParameterHash' from a list of 'PublicParameter's.
+--
+-- In the context of verification the public parameters are just the
+-- concatenation of the bytes that the program outputs.
+--
+-- The digest is computed as SHA256 hash with the first three bits set to 0.
+--
+hashPublicParameters :: Foldable f => f PublicParameter -> PublicParameterHash
+hashPublicParameters l = PublicParameterHash
+    $ BS.toShort
+    $ clearBits
+    $ H.finalize
+    $ H.updates H.init
+    $ fmap (BS.fromShort . _publicParameter)
+    $ toList
+    $ l
+  where
+    clearBits x = B.cons (0x1f .&. B.head x) (B.tail x)
+
+-- -------------------------------------------------------------------------- --
+-- Plonk Verifier API
+
+-- | Verify the claim that the program with the given id was invoked with the
+-- given list of public parameters.
+--
+verifyPlonkBn254
+    :: VKey
+        -- ^ The verifying key in bytes. This key represents the particular
+        -- version of the verifier. It must match the respective version of the
+        -- prover that was used to generate the proof. Otherwise verification
+        -- fails.
+    -> Proof
+        -- ^ The proof object for the invocation of program with the respective
+        -- public parameters.
+    -> ProgramId
+        -- ^ The program identifier. A program is valid only in the context of a
+        -- particular verifying key. It also the number and types of public
+        -- parameters are well defined.
+    -> [PublicParameter]
```

Comment:
This is problematic. The verifier just concatenates all parameters to a single byte string. The proof does not constrain how the parameters are split. So, different lists are all valid. It is therefore up to the user to define and validate the encoding of the parameters as a whole.